crates/chia-bls/src/cached_bls.rs

```diff
@@ -1,253 +1,260 @@
-// to see them again unless there's a reorg. However, a spend in the mempool
-// is likely to reappear in a block later, so we can save having to do the pairing
-// again. So caching is primarily useful when synced and monitoring the mempool in real-time.
-
-use crate::aggregate_verify_gt as agg_ver_gt;
-use crate::gtelement::GTElement;
-use crate::hash_to_g2;
-use crate::PublicKey;
-use crate::Signature;
-use lru::LruCache;
-
-use sha2::{Digest, Sha256};
 use std::borrow::Borrow;
 use std::num::NonZeroUsize;
 
-#[cfg(feature = "py-bindings")]
-use pyo3::exceptions::PyValueError;
-#[cfg(feature = "py-bindings")]
-use pyo3::pybacked::PyBackedBytes;
-#[cfg(feature = "py-bindings")]
-use pyo3::types::{PyAnyMethods, PyList};
-#[cfg(feature = "py-bindings")]
-use pyo3::{pyclass, pymethods, PyResult};
+use lru::LruCache;
+use sha2::{Digest, Sha256};
 
-#[cfg_attr(feature = "py-bindings", pyclass(name = "BLSCache"))]
-#[derive(Clone)]
-pub struct BLSCache {
-    cache: LruCache<[u8; 32], GTElement>, // LRUCache of hash(pubkey + message) -> GTElement
+use crate::{aggregate_verify_gt, hash_to_g2};
+use crate::{GTElement, PublicKey, Signature};
+
+/// This cache is a bit weird because it's trying to account for validating
+/// mempool signatures versus block signatures. When validating block signatures,
+/// there's not much point in caching the pairings because we're probably not going
+/// to see them again unless there's a reorg. However, a spend in the mempool
+/// is likely to reappear in a block later, so we can save having to do the pairing
+/// again. So caching is primarily useful when synced and monitoring the mempool in real-time.
+#[cfg_attr(feature = "py-bindings", pyo3::pyclass(name = "BLSCache"))]
+#[derive(Debug, Clone)]
+pub struct BlsCache {
+    // sha256(pubkey + message) -> GTElement
+    cache: LruCache<[u8; 32], GTElement>,
 }
 
-impl Default for BLSCache {
+impl Default for BlsCache {
     fn default() -> Self {
-        Self::new(None)
+        Self::new(NonZeroUsize::new(50000).unwrap())
     }
 }
 
-impl BLSCache {
-    pub fn new(cache_size: Option<NonZeroUsize>) -> BLSCache {
-        let cache: LruCache<[u8; 32], GTElement> = LruCache::new(
-            cache_size.unwrap_or(NonZeroUsize::new(50000).expect("50000 should be non-zero")),
-        );
-        Self { cache }
+impl BlsCache {
+    pub fn new(cache_size: NonZeroUsize) -> Self {
+        Self {
+            cache: LruCache::new(cache_size),
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.cache.len()
     }
 
-    pub fn aggregate_verify<
-        M: IntoIterator<Item = T>,
-        T: AsRef<[u8]>,
-        P: IntoIterator<Item = U>,
-        U: Borrow<PublicKey>,
-    >(
+    pub fn is_empty(&self) -> bool {
+        self.cache.is_empty()
+    }
+
+    pub fn aggregate_verify(
         &mut self,
-        pks: P,
-        msgs: M,
+        pks: impl IntoIterator<Item = impl Borrow<PublicKey>>,
```

Comment:
Personal preference, `impl Trait` is syntax sugar for generics, so this doesn't change behavior at all.